chain-signatures/node/src/protocol/triple.rs

```diff
@@ -35,52 +44,204 @@ pub struct Triple {
     pub public: TriplePub<Secp256k1>,
 }
 
+#[derive(Clone)]
 pub struct TripleGenerator {
     pub id: TripleId,
     pub participants: Vec<Participant>,
-    pub protocol: TripleProtocol,
-    pub timestamp: Option<Instant>,
+    pub protocol: Arc<TripleProtocol>,
+    pub timestamp: Arc<RwLock<Option<Instant>>>,
     pub timeout: Duration,
 }
 
 impl TripleGenerator {
     pub fn new(
+        me: Participant,
+        threshold: usize,
         id: TripleId,
         participants: Vec<Participant>,
-        protocol: TripleProtocol,
         timeout: u64,
-    ) -> Self {
-        Self {
+    ) -> Result<Self, InitializationError> {
+        let protocol = Arc::new(RwLock::new(
+            cait_sith::triples::generate_triple::<Secp256k1>(&participants, me, threshold)?,
+        ));
+
+        Ok(Self {
             id,
             participants,
             protocol,
-            timestamp: None,
+            timestamp: Arc::new(RwLock::new(None)),
             timeout: Duration::from_millis(timeout),
+        })
+    }
+
+    pub async fn message(&self, from: Participant, data: MessageData) {
+        let mut protocol = self.protocol.write().await;
+        protocol.message(from, data);
+    }
+
+    pub async fn messages(&self, from: Participant, data: Vec<MessageData>) {
+        let mut protocol = self.protocol.write().await;
+        for data in data {
+            protocol.message(from, data);
         }
     }
 
-    pub fn poke(&mut self) -> Result<Action<TripleGenerationOutput<Secp256k1>>, ProtocolError> {
-        let timestamp = self.timestamp.get_or_insert_with(Instant::now);
-        if timestamp.elapsed() > self.timeout {
-            tracing::warn!(
-                id = self.id,
-                elapsed = ?timestamp.elapsed(),
-                "triple protocol timed out"
-            );
+    pub fn spawn_execution(
+        &self,
+        me: Participant,
+        my_account_id: &AccountId,
+        epoch: u64,
+    ) -> JoinHandle<GeneratorOutcome> {
+        tokio::task::spawn({
+            let mut generator = self.clone();
+            let my_account_id = my_account_id.clone();
+            async move { generator.execute(me, &my_account_id, epoch).await }
+        })
+    }
```

Comment:
This looks better, at first glance what we do in `TripleGenerator::execute`, I think we get real parallelism here.